### PR TITLE
Make default resolvers subclassable

### PR DESCRIPTION
**Pull request by pauxus**
Implementation for #1567.

This is simply an aggregation of #1568, #1569, #1570, #1581. Since these are all related, it IMHO makes more sense to really merge them all together.

### CODE REVIEW

File: flyway-core/src/main/java/org/flywaydb/core/internal/resolver/sql/SqlMigrationExecutor.java [hunk 1]
```diff
@@ -50,18 +50,19 @@
...
-        JdbcTemplate jdbcTemplate = connection == dbSupport.getJdbcTemplate().getConnection()
-                ? dbSupport.getJdbcTemplate()
-                : new JdbcTemplate(connection, 0);
+        // TODO: This reverts parts of commit ca87e59aed534e4ad9bef610517fc595cf2c39c8
+//        JdbcTemplate jdbcTemplate = connection == dbSupport.getJdbcTemplate().getConnection()
+//                ? dbSupport.getJdbcTemplate()
+//                : new JdbcTemplate(connection, 0);
 
-        getSqlScript().execute(jdbcTemplate);
+        getSqlScript().execute(new JdbcTemplate(connection, 0));
```
Comment (pauxus) [inline flyway-core/src/main/java/org/flywaydb/core/internal/resolver/sql/SqlMigrationExecutor.java]: I needed to revert this change, so please take a special look here

**Comment by axelfontaine** [comment]
Closing old pull request. Too many conflicts by now and too large to be safely merged in the first place anyway.
